src/libs/print_settings.c

```diff
@@ -0,0 +1,1398 @@
+  dat.style_append = dt_conf_get_bool("plugins/print/print/style_append");
+  dat.bpp = *printer_profile ? 16 : 8; // set to 16bit when a profile is to be applied
+  dat.ps = ps;
+
+  char* style = dt_conf_get_string("plugins/print/print/style");
+  if (style)
+  {
+    g_strlcpy(dat.style, style, sizeof(dat.style));
+    g_free(style);
+  }
+
+  // the flags are: ignore exif, display byteorder, high quality, thumbnail
+  const int high_quality = 1;
+  dt_imageio_export_with_flags
+    (imgid, "unused", &buf, (dt_imageio_module_data_t *)&dat, 1, 0, high_quality, 0, NULL, FALSE, 0, 0, 1, 1);
+
+  // after exporting we know the real size of the image, compute the layout
+
+  // compute print-area (in inches)
+  int32_t px=0, py=0, pwidth=0, pheight=0;
+  int32_t ax=0, ay=0, awidth=0, aheight=0;
+  int32_t ix=0, iy=0, iwidth=dat.width, iheight=dat.height;
+
+  dt_get_print_layout (imgid, &ps->prt, width_pix, height_pix,
+                       &px, &py, &pwidth, &pheight,
+                       &ax, &ay, &awidth, &aheight,
+                       &ix, &iy, &iwidth, &iheight);
+
+  const int margin_top    = iy;
+  const int margin_left   = ix;
+  const int margin_right  = pwidth - iwidth - ix;
+  const int margin_bottom = pheight - iheight - iy;
+
+  if (darktable.unmuted & DT_DEBUG_PRINT)
+    printf("[print] margins top %d ; bottom %d ; left %d ; right %d\n",
+            margin_top, margin_bottom, margin_left, margin_right);
+
+  // we have the exported buffer, let's apply the printer profile
+
+  if (*printer_profile)
+    if (dt_apply_printer_profile(imgid, (void **)&(dat.ps->buf), dat.width, dat.height, dat.bpp, printer_profile, pintent))
+    {
+      dt_control_log(_("cannot apply printer profile `%s'"), printer_profile);
+      dt_control_queue_redraw();
+      return;
+    }
+
+  float border;
+  dt_pdf_parse_length("0 mm", &border);
+
+  const float page_width  = dt_pdf_mm_to_point(width);
+  const float page_height = dt_pdf_mm_to_point(height);
+
+  char filename[PATH_MAX] = { 0 };
+  snprintf(filename, sizeof(filename), "/tmp/pf_%d.pdf", imgid);
+
+  const int icc_id = 0;
+
+  dt_pdf_t *pdf = dt_pdf_start(filename, page_width, page_height, ps->prt.printer.resolution, DT_PDF_STREAM_ENCODER_FLATE);
+
+/*
+  // ??? should a profile be embedded here?
+  if (*printer_profile)
+    icc_id = dt_pdf_add_icc(pdf, printer_profile);
+*/
+  dt_pdf_image_t *pdf_image = dt_pdf_add_image(pdf, (uint8_t *)dat.ps->buf, dat.width, dat.height, 8, icc_id, border);
+
+  //  PDF bounding-box has origin on bottom-left
+  pdf_image->bb_x      = dt_pdf_pixel_to_point((float)margin_left, ps->prt.printer.resolution);
+  pdf_image->bb_y      = dt_pdf_pixel_to_point((float)margin_bottom, ps->prt.printer.resolution);
+  pdf_image->bb_width  = dt_pdf_pixel_to_point((float)iwidth, ps->prt.printer.resolution);
+  pdf_image->bb_height = dt_pdf_pixel_to_point((float)iheight, ps->prt.printer.resolution);
+
+  if (ps->prt.page.landscape && (dat.width > dat.height))
```

Comment:
That's fine, the image is always on the same orientation on screen. I was puzzled when I implemented that but it is correct.